endpoints-control/src/main/java/com/google/api/control/ConfigFilter.java

```diff
@@ -171,8 +176,38 @@ public static ReportingRule getReportRule(ServletRequest req) {
     return (MethodRegistry.Info) httpRequest.getAttribute(METHOD_INFO_ATTRIBUTE);
   }
 
+  /**
+   * Get the "real" HTTP method for the request, taking into account the X-HTTP-Method-Override
+   * header.
+   *
+   * @param req a {@code ServletRequest}
+   * @return the HTTP method for the request
+   */
+  public static String getRealHttpMethod(ServletRequest req) {
+    HttpServletRequest httpRequest = (HttpServletRequest) req;
+    String method = (String) httpRequest.getAttribute(HTTP_METHOD_OVERRIDE_ATTRIBUTE);
+    if (method != null) {
+      return method;
+    }
+    return httpRequest.getMethod();
+  }
+
   @Override
   public void destroy() {
     // unused
   }
+
+  @VisibleForTesting
+  static String getRequestMethodOverride(HttpServletRequest request) {
+    Enumeration headerNames = request.getHeaderNames();
+    String methodOverride = null;
+    while (headerNames.hasMoreElements()) {
+      String headerName = (String) headerNames.nextElement();
+      if (headerName.toLowerCase().equals("x-http-method-override")) {
+        methodOverride = request.getHeader(headerName);
+        break;
+      }
+    }
+    return methodOverride != null ? methodOverride.toUpperCase() : null;
```

Comment:
In a lot of languages, this is where you'd use an Optional. Is that not recommended in Java?

Reply:
I don't think this does us much good. Since it is a dynamic property, using Optional would mean getRealHttpMethod has to handle 3 cases instead of 2, as it still has to check nullness.